### PR TITLE
[Keyboard] Add Yugo-M Controller

### PR DESCRIPTION
**Pull request by tomic1785**
## Description

Yugo-M is an ARM-based replacement controller for IBM Model M keyboards. 
To my knowledge, there aren't any products of the same kind available for purchase at this time, which are supported by QMK Firmware. 
Check out the [project repo](https://github.com/tomic1785/Yugo-M-controller-project)
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/yugo_m/model_m_101/config.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#define MANUFACTURER    IBM
+#define PRODUCT         Model M
+#define DESCRIPTION     QMK-powered IBM Model M
+
+/* Address for jumping to bootloader on STM32 chips. */
+/* It is chip dependent, the correct number can be looked up here:
+ * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
+ */
+#define STM32_BOOTLOADER_ADDRESS 0x1FFFD800
+
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/config.h:35]:
```suggestion

```

File: keyboards/yugo_m/model_m_101/model_m_101.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+void matrix_init_kb(void) {
+    // put your keyboard start-up code here
+    // runs once when the firmware starts up
+    matrix_init_user();
+    led_init_ports();
+}
+
+void led_init_ports() {
+    // Set our LED pins as output
+    palSetPadMode(GPIOA, 2, PAL_MODE_OUTPUT_PUSHPULL); // NumLock LED
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/model_m_101.c:27]: Needs to be updated to <https://docs.qmk.fm/#/internals_gpio_control>

File: keyboards/yugo_m/model_m_101/model_m_101.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+// #ifndef MODELM101_H
+// #define MODELM101_H
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/model_m_101.h:17]:
```suggestion
#pragma once
```

File: keyboards/yugo_m/model_m_101/rules.mk [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#
+BACKLIGHT_ENABLE = no
+BOOTMAGIC_ENABLE = no	  # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = no	  # Mouse keys
+EXTRAKEY_ENABLE = yes	  # Audio control and System control
+CONSOLE_ENABLE = no	  # Console for debug
+COMMAND_ENABLE = no     # Commands for debug and configuration
+NKRO_ENABLE = no	      # USB Nkey Rollover
+AUDIO_ENABLE = no
+
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/rules.mk:15]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/yugo_m/model_m_101/model_m_101.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    palClearPad(GPIOA, 2);
+    palSetPadMode(GPIOA, 1, PAL_MODE_OUTPUT_PUSHPULL); // CapsLock LED
+    palClearPad(GPIOA, 1);
+    palSetPadMode(GPIOA, 0, PAL_MODE_OUTPUT_PUSHPULL); // ScrollLock LED
+    palClearPad(GPIOA, 0);
+}
+
+void led_set_kb(uint8_t usb_led) {
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/model_m_101.c:35]: Needs to be updated to <https://docs.qmk.fm/#/custom_quantum_functions?id=example-led_update_kb-implementation>

**Review by zvecr — CHANGES_REQUESTED** [review]
Also the halconf,chconf,mcuconf should follow the new strategy per <https://github.com/qmk/qmk_firmware/tree/master/keyboards/planck/rev6>. Where only the required differences are changed, and otherwise rely on the absence of the file to use the defaults.

File: keyboards/yugo_m/model_m_101/model_m_101.c (hunk 2 above)
Comment (tomic1785) [reply keyboards/yugo_m/model_m_101/model_m_101.c]: Done. I hopefully got it right. Tested on a keyboard, doesn't break anything.

File: keyboards/yugo_m/model_m_101/model_m_101.c (hunk 5 above)
Comment (tomic1785) [reply keyboards/yugo_m/model_m_101/model_m_101.c]: Also done. Thank you for help!

File: keyboards/yugo_m/model_m_101/config.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x0000
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    IBM
+#define PRODUCT         Model M
```
Comment (noroadsleft) [inline keyboards/yugo_m/model_m_101/config.h:27]:
```suggestion
#define PRODUCT         Yugo-M (Model M replacement controller)
```

File: keyboards/yugo_m/model_m_101/config.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x0000
```
Comment (noroadsleft) [inline keyboards/yugo_m/model_m_101/config.h:24]:
```suggestion
#define PRODUCT_ID      0x8E81  // Yugo-M
```

`echo -n 'Yugo-M' | shasum | cut -c1-4`

File: keyboards/yugo_m/model_m_101/model_m_101.c [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  setPinOutput(A1);
+  setPinOutput(A0);
+}
+
+void matrix_init_kb(void) {
+    // put your keyboard start-up code here
+    // runs once when the firmware starts up
+    matrix_init_user();
+}
+
```
Comment (noroadsleft) [inline keyboards/yugo_m/model_m_101/model_m_101.c:32]: This is the same as the default function built into QMK, so there's no need to include it here.

```suggestion
```

File: keyboards/yugo_m/readme.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# Yugo-M
+#### The stm32-based IBM Model M controller
```
Comment (noroadsleft) [inline keyboards/yugo_m/readme.md:2]:
```suggestion
# Yugo-M - An STM32-based IBM Model M controller
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

File: keyboards/yugo_m/model_m_101/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    [_CL] = LAYOUT_all(
+      _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,   _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,   _______, _______, _______,   _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,   _______, _______, _______,   _______, _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,                                _______, _______, _______,
+      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,            _______,            _______, _______, _______, _______,
+      _______,          _______,                            _______,                                     _______,          _______,   _______, _______, _______,   _______,          _______),
```
Comment (fauxpark) [inline keyboards/yugo_m/model_m_101/keymaps/default/keymap.c:60]:
```suggestion
```

File: keyboards/yugo_m/model_m_101/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+enum keyboard_layers {
+    _BL, // Base Layer
+    _FL, // Function Layer
+    _CL  // Control Layer
```
Comment (fauxpark) [inline keyboards/yugo_m/model_m_101/keymaps/default/keymap.c:21]:
```suggestion
```

File: keyboards/yugo_m/model_m_101/model_m_101.c [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+bool led_update_kb(led_t led_state) {
+    bool res = led_update_user(led_state);
+    if(res) {
+        // writePin sets the pin high for 1 and low for 0.
+        // In this example the pins are inverted, setting
+        // it low/0 turns it on, and high/1 turns the LED off.
+        // This behavior depends on whether the LED is between the pin
+        // and VCC or the pin and GND.
```
Comment (fauxpark) [inline keyboards/yugo_m/model_m_101/model_m_101.c:34]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/yugo_m/model_m_101/model_m_101.c [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+#include "model_m_101.h"
+
+void keyboard_pre_init_user(void) {
+  // Call the keyboard pre init code.
+
+  // Set our LED pins as output
+  setPinOutput(A2);
+  setPinOutput(A1);
+  setPinOutput(A0);
+}
```
Comment (zvecr) [inline keyboards/yugo_m/model_m_101/model_m_101.c:25]:
```suggestion
void keyboard_pre_init_kb(void) {
  // Set our LED pins as output
  setPinOutput(A2);
  setPinOutput(A1);
  setPinOutput(A0);
  
  keyboard_pre_init_user();
}
```

**Review by zvecr — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Pull request merged**